onnxruntime/core/providers/mkldnn/nn/batch_norm.cc

```diff
@@ -0,0 +1,287 @@
+// Copyright(C) 2018 Intel Corporation
+// Licensed under the MIT License
+
+#ifdef _WIN32
+#pragma warning(disable : 4244)
+#endif
+
+#include "core/providers/mkldnn/mkldnn_common.h"
+#include "core/providers/mkldnn/nn/batch_norm.h"
+#include "core/providers/mkldnn/mkldnn_fwd.h"
+#include "core/providers/mkldnn/memcpy_s.h"
+#include "core/providers/cpu/nn/batch_norm_helper.h"
+
+namespace onnxruntime {
+namespace mkl_dnn {
+
+namespace {
+// Struct which encapsulates parameters for MKLDNN BatchNorm primitive.
+struct BatchNormParams {
+  const mkldnn::memory::dims& src_dims;
+  const mkldnn::memory::dims& scale_dims;
+  const mkldnn::memory::dims& b_dims;
+  const mkldnn::memory::dims& mean_dims;
+  const mkldnn::memory::dims& var_dims;
+  const mkldnn::memory::dims& dst_dims;
+  const float epsilon;
+  const int num_dimensions;
+
+  BatchNormParams(const mkldnn::memory::dims& src_dims_mkl,
+    const mkldnn::memory::dims& scale_dims_mkl,
+    const mkldnn::memory::dims& b_dims_mkl, const mkldnn::memory::dims& mean_dims_mkl,
+    const mkldnn::memory::dims& var_dims_mkl, const mkldnn::memory::dims& dst_dims_mkl,
+     float eps, int dimensions)
+    : src_dims(src_dims_mkl),
+    scale_dims(scale_dims_mkl),
+    b_dims(b_dims_mkl),
+    mean_dims(mean_dims_mkl),
+    var_dims(var_dims_mkl),
+    dst_dims(dst_dims_mkl),
+    epsilon(eps),
+    num_dimensions(dimensions) {}
+
+  // Used as the key for BatchNorm Primitive Reuse Pool.
+  std::string ToString() const {
+    std::string key;
+    key.reserve(128);
+    key.append("BatchNorm_");
+    AddDimsToKey(key, src_dims);
+    AddDimsToKey(key, scale_dims);
+    AddDimsToKey(key, b_dims);
+    AddDimsToKey(key, mean_dims);
+    AddDimsToKey(key, var_dims);
+    AddDimsToKey(key, dst_dims);
+    return key;
+  }
+};
+
+template <typename T>
+class BatchNormPrimitive final : public PrimitiveBase {
+ public:
+  explicit BatchNormPrimitive(const BatchNormParams& params)
+      : cpu_engine_(GetEngine()) {
+    context_.stream.reset(new mkldnn::stream(mkldnn::stream::kind::eager));
+    if (context_.batchnorm_fwd == nullptr) {
+      Initialize(params);
+    }
+  }
+
+  ~BatchNormPrimitive() = default;
+
+  void Compute(const T* src_data, const T* scale_data, const T* b_data, 
```

Comment:
please follow similar guidelines as Relu. data that is intended to be mutated (like output data) should not be passed as const to begin with (to avoid const_cast later on).